### ISSUE TITLE
optimize,diff/fd: impedance mismatch in Hessian and mat.MutableSymmetric

### ISSUE THREAD

**Issue by sbinet** [gonum/gonum#223]
recently, I wanted to numerically compute a hessian, so I used:
https://godoc.org/gonum.org/v1/gonum/diff/fd#Hessian
```go
// Hessian approximates the Hessian matrix of the multivariate function f
// at the location x. That is
//  H_{i,j} = ∂^2 f(x)/∂x_i ∂x_j
// If dst is not nil, the resulting H will be stored in-place into dst
// and returned, otherwise a new matrix will be allocated first. Finite difference
// formula and other options are specified by settings. If settings is nil,
// the Hessian will be estimated using the Forward formula and a default step size.
//
// Hessian panics if the size of dst and x is not equal, or if the derivative
// order of the formula is not 1.
func Hessian(
    dst *mat.SymDense, f func(x []float64) float64, 
    x []float64, settings *Settings,
) *mat.SymDense
```

so I could feed that into `optimize.Problem.Hess`:
https://godoc.org/gonum.org/v1/gonum/optimize#Problem
```go
type Problem struct {
    // Func evaluates the objective function at the given location. Func
    // must not modify x.
    Func func(x []float64) float64

    // Grad evaluates the gradient at x and stores the result in-place in grad.
    // Grad must not modify x.
    Grad func(grad []float64, x []float64)

    // Hess evaluates the Hessian at x and stores the result in-place in hess.
    // Hess must not modify x.
    Hess func(hess mat.MutableSymmetric, x []float64)

    // Status reports the status of the objective function being optimized and any
    // error. This can be used to terminate early, for example when the function is
    // not able to evaluate itself. The user can use one of the pre-provided Status
    // constants, or may call NewStatus to create a custom Status value.
    Status func() (Status, error)
}
```

so, I had to resort to:

```go
p.Hess = func(hess mat.MutableSymmetric, x []float64) {
	fd.Hessian(hess.(*mat.SymDense), f.fct, x, nil)
}
```

looking at `fd.Hessian`, it seems to me the only requirement for using `*mat.SymDense` is coming from the ability to pass `nil` as an optional parameter (as a Hessian matrix) and return it (as an output value.)

couldn't we just use `mat.MutableSymmetric` everywhere ?

**Comment by btracey** [comment]
That's not the motivation for returning `*SymDense`. The motivation is that `fd.Hessian` has no way except to populate a `Dense` matrix (it has to compute the finite difference in all the dimensions to perform the documented behavior). It's good policy to return specific types that are generated.

I'm sympathetic to the motivation, but why should `fd.Hessian` return a `MutableSymmetric` as opposed to any number of possible interfaces? The only real reason is to match with `optimize`, but fixing on that particular interface seems like it will cause other frictions down the road.

**Comment by sbinet** [comment]
yes, that's why I was (perhaps in a very convoluted way) alluding to not returning any value for `diff/fd.Hessian`.

were `diff/fd.Hessian` not returning any value, then no need to assume any concrete matrix type, just take `mat.MutableSymmetric` as a destination and use that interface to do its work.

**Comment by kortschak** [comment]
Given https://github.com/gonum/gonum/issues/705#issuecomment-478221863, why not make this `Hessian(dst *mat.SymDense, f func(x []float64) float64, x []float64, settings *Settings)` with the requirement that `dst` be zero-sized or correctly sized.

**Comment by btracey** [comment]
Yes, I agree. I think this would always require a shim function because `fd.Hessian` needs a `*Settings` while `optimize.Problem.Hess` will not. But, as per the impedance mismatch, that can be reduced.

We could reduce it further by changing Problem.Hess to take in a `*SymDense` also. Then this would look like

```
problem.Func = f
problem.Grad = func(g, x []float64) []float64{
     return fd.Gradient(g, f, x, gradSettings)
}
problem.Hess = func(s *SymDense, x) {
    fd.Hessian(s, f, x, hessSettings)
}
```
which is as close to a match as possible.

I originally thought this was bad because it constrains `optimize` to require a `*SymDense` for the Hessian representation, which, for example, makes it impossible to use a sparse Hessian. However, my recent failed efforts at making `Cholesky` abstract, plus @vladimir-ch pointing out it's impossible to make the current implementation of `Newton` abstract, make it seem like that goal is misguided. If that goal is misguided, then we should be making everything as easy as possible to use.

An alternative is to abstract the `optimize` backend to allow for different front-ends (as suggested in #925 ), rather than try to have optimize be all things for all people. Then an ouside user could define a `sparsenewton` package, with a different definition of `problem.Hess` that meets the requirements needed for that particular sparse solver.

The downside to changing `problem.Hess` to a `*SymDense` is that it means _every_ `optimize.Method`, even those defined externally, have to use the dense representation. On the other hand, `optimize.Newton` is basically the simplest possible algorithm using a Hessian, and so if that method can't be abstracted, it's not clear any method can be abstracted safely. You can currently do things not-so-safely, by guaranteeing the return from problem.Hess matches the interface the method actually needs and then use type assertions. This is ripe for error, however, and it seems like solving #925 also solves this problem (user defines their own Problem instead of shoe-horning through optimize).

In short: 1) We implement the change suggested here by @kortschak  2) We change  `optimize` to take `*SymDense`

Does that make sense @sbinet @vladimir-ch @kortschak ? (Also adding some comments to #354 )

I apologize for not understanding earlier and making stability difficult.
